Accept statsd from authenticated TLS connections

evanj opened the pull request:
#### Summary

By default, Veneur (and statsd, and dogstatsd) accepts stats from any
client that can send UDP packets to the correct port. For slightly
bizarre reasons, it is useful to only accept data from authenticated
clients. An "easy" way to do this securely is to use an encrypted
TLS connection, and require the client to present a certificate
signed by a trusted authority. This reuses the existing
authentication mechanism in TLS, which is hopefully secure.

This adds configuration options to read the statsd protocol from a
TLS connection. It will only accept connections with a client
certificate, signed by the configured authority.


#### Motivation

See Issue #117 


#### Test plan

Currently we (Bluecore Inc) are running this change as an experiment for getting metrics out of App Engine. We will be ramping up our usage of it over the next couple weeks.

There are no unit tests here yet: If this seems reasonable, I'll attempt to add some for the TCP connection reader, and possibly for the TLS listen socket setup.


#### Rollout/monitoring/revert plan

This adds `tls.connects` and `tls.disconnects` counters to monitor connections with TLS. Maybe this should be a gauge, but then you might not see if clients are in reconnect loops.

evanj commented:
Again, I totally understand if it doesn't make sense to accept this. However, I'm also happy to add tests or make whatever changes you want if you are willing to accept it, since it would simplify our lives a bit to not maintain a private fork. Thanks!

evanj commented:
Great, I'll take a look at adding unit tests to this. And yes, it would 100% be possible to change the configuration option to be `tcp_listener` so it would read statsd requests from a TCP socket, without using TLS at all. I'm happy to change this to do that. My rough plan is that there would effectively be three modes:

1. Raw TCP: `tcp_address` is set, but no keys are set: Listen on the TCP port for unauthenticated and unencrypted requests.
2. Encryption only:`tcp_address` is set, with `tls_key` and `tls_certificate`: Listen for *TLS* connections, for *encrypted data* only.
3. Encryption and client authentication: `tcp_address` is set, with `tls_key`, `tls_certificate`, `tls_authority_certificate`: Require a client certificate.

cory-stripe commented:
That sounds great, looking forward to the results!

evanj commented:
I've made a fairly substantial update. This now supports TCP, encrypted TLS, or encrypted and authenticated TLS connections. I also added a unit test to check all these configurations.

The unit test builds on PR #119 so this pull request currently looks enormous. Once someone makes a merge/no merge decision on that PR, I'll update this request appropriately. However, the non-test parts are independently, and can be reviewed in parallel.

Thanks again for considering this! I'm aware that this is probably fairly esoteric, and we might be the only people in the universe who actually use this feature.

cory-stripe commented on README.md:
````diff
@@ -237,6 +237,55 @@ sysctl -w net.ipv4.udp_mem="4648512 6198016 9297024"
+
+```
+tcp_address: "localhost:8129"
+tls_certificate: |
````
Is there a reason you chose to use a string here rather than a file on disk? Could we support both? (Genuine question, I don't know of a reason for either… but I know we might use keys/certs we already have on disk for other things?)

evanj replied:
Not any particularly good reason, other than it was easy to embed it. If you would prefer, I'm happy to make this a path.

cory-stripe replied:
I don't feel strongly either way. Since we're not using it and you are, I think what's easier for you is fine for now.

If someone wants to change it in the future, it's a pretty small change. Onward! :)

cory-stripe replied:
👍

cory-stripe commented on cmd/veneur/main.go:
```diff
@@ -71,6 +71,14 @@ func main() {
+		defer func() {
+			server.ConsumePanic(recover())
+		}()
+		server.AcceptTCP(server.shutdownChan)
```
As is done with the trace reader above, can we limit this to when the config says it should run?

evanj replied:
Great suggestion.

cory-stripe commented on server.go:
```diff
@@ -166,8 +175,47 @@ func NewFromConfig(conf Config) (ret Server, err error) {
 	ret.HTTPAddr = conf.HTTPAddress
 	ret.ForwardAddr = conf.ForwardAddress
 
+	ret.shutdownChan = make(chan struct{})
```
As with my other common about config, we want to allow the user the option of not running this by default. Can we skip this bit if there's no TcpAddress?

evanj replied:
Agreed.

cory-stripe commented on server.go:
```diff
@@ -376,6 +424,114 @@ func (s *Server) ReadTraceSocket(packetPool *sync.Pool, reuseport bool) {
+		if !tlsConn.ConnectionState().HandshakeComplete {
+			err := tlsConn.Handshake()
+			if err != nil {
+				log.WithError(err).Error("TLS Handshake failed")
```
Should we have a metric here as well?

evanj replied:
I've added `tcp.tls_handshake_errors` here if that makes sense.

cory-stripe commented on worker.go:
```diff
@@ -138,6 +139,7 @@ func (w *Worker) Work() {
 	for {
 		select {
 		case m := <-w.PacketChan:
+			fmt.Println("WTF ProcessMetric", m)
```
Probably didn't mean to leave that, lol. 😸

evanj replied:
Nope definitely not. Thanks! :) I thought I searched for all `'fmt` calls, but clearly I didn't check the diff carefully enough, thanks. :)

cory-stripe commented on worker.go:
```diff
@@ -241,6 +243,8 @@ func (w *Worker) Flush() WorkerMetrics {
 	processed := w.processed
 	imported := w.imported
 
+	fmt.Println("Worker.Flush()", processed, imported)
```
Same here!

cory-stripe commented:
This is looking really great, @evanj! I had a few comments, questions and fixes in that review. Thank you for being so responsive and thoughtful about your change!

evanj commented:
Hello! This is definitely not critical, but just wanted to clarify that I *think* I've addressed all the above concerns. Ideally IMO, someone should approve #119, then I can rebase this change against the new master, just to separate out the test refactoring. Or you could just merge this one since it includes both. Thanks!

cory-stripe commented:
Shit, I'm sorry @evanj. I had a pending couple questions, added. Also @aditya-stripe will check this out soonish and we should be good to go.

Sorry about that!

ChimeraCoder commented on server.go:
```diff
@@ -376,6 +426,107 @@ func (s *Server) ReadTraceSocket(packetPool *sync.Pool, reuseport bool) {
+			if err != nil {
+				// TODO: When does this happen? Is there a better way to handle these errors?
+				log.WithError(err).Error("Error accepting TCP connection; accept exiting")
+				return
```
In this case, the TCP listener will termiate, but the program will continue running. Is that what you want? As opposed to either continuing to handle the next connection or panicking (triggering a restart)?

evanj replied:
I'm not totally sure what the right answer is here. Part of the problem is this error happens when shutting down, since another goroutine closes the listener. Other then that, I'm not sure when this can happen (`man 2 accept` suggests that "out of file descriptors" is another possible case)

I'm not sure looping and retrying again makes sense, because it could hit an infinite error loop if the error is permanent.

I've implemented what is maybe a better alternative: I've caught the "operation on closed socket" that happens when trying to cleanly shutdown, just logging and returning in that case. It now will crash on any other error. Thoughts?

ChimeraCoder commented on cmd/veneur/main.go:
```diff
@@ -71,6 +71,16 @@ func main() {
 		}
 	}()
 
+	// Accept connections on the TCP port if enabled
```
I think you'll want this in a separate goroutine (similar to the trace reader), so it can start asynchronously.

evanj replied:
Sure. I had to rearrange the code that starts the listener to make it work with shutting down (used in the unit tests).

ChimeraCoder commented on server.go:
```diff
@@ -376,6 +426,107 @@ func (s *Server) ReadTraceSocket(packetPool *sync.Pool, reuseport bool) {
+	log.WithField("peer", conn.RemoteAddr().String()).Debug("Closing TCP connection")
+}
+
+// StartTCP listens on Server.TCPAddr for new connections, starting a goroutine for each.
```
Can we update the name of this to be similar to the other `Read...` methods we have for metrics and trace? The relationship isn't really 1:1, but they seem to be pretty analogous.

ChimeraCoder commented on server.go:
```diff
@@ -376,6 +426,107 @@ func (s *Server) ReadTraceSocket(packetPool *sync.Pool, reuseport bool) {
+				logrus.ErrorKey: err,
+				"packet":        strconv.QuoteToASCII(buf.Text()),
+			}).Error("Could not parse packet; Closing TCP connection")
+			s.statsd.Count("packet.error_total", 1, []string{"packet_type:metric"}, 1.0)
```
Just to be clear: this is intended to listen only for metrics (and not events or service checks)?

evanj replied:
Correct, although mostly just because statsd is all I needed. Now that I look at this: It should be pretty trivial to change this to call `HandleMetricPacket` for each line, which actually will reduce duplication, as well as support events and service checks. I've made this change. Thanks!

ChimeraCoder commented on README.md:
```diff
@@ -237,6 +237,55 @@ sysctl -w net.ipv4.udp_mem="4648512 6198016 9297024"
+## TCP connections
+
+Veneur supports reading the statds protocol from TCP connections. This is mostly to support TLS encryption and authentication, but might be useful on its own. Since TCP is a continuous stream of bytes, this requires each stat to be terminated by a new line character ('\n'). Most statsd clients only add new lines between stats within a single UDP packet, and omit the final trailing new line. This  ...
+
```
If this is only going to listen to metrics (see other comment), we should mention that here as well.

evanj replied:
Fixed!

evanj commented on example.yaml:
```diff
@@ -40,3 +40,13 @@ aws_access_key_id: ""
+# Listen address for statsd over TCP
+tcp_address: ""
+
+# TLS server private key and certificate for encryption (must specify both)
```
Good suggestion done.

evanj commented on server.go:
```diff
@@ -166,8 +175,48 @@ func NewFromConfig(conf Config) (ret Server, err error) {
+	if conf.TcpAddress != "" {
+		ret.TCPAddr, err = net.ResolveTCPAddr("tcp", conf.TcpAddress)
+		if err != nil {
+			return
```
This function seems to be used "named returns". I followed the pattern that already exists here. For example, see how errors are handled when resolving the UDP address, directly above here on line 171

evanj commented:
I rebased this change to avoid conflicts in `server_test.go`; sorry that messes up code review a bit.

"Major" changes:
* This now calls `HandleMetricPacket` to handle all packet types, which also reduces duplication.
* Starting the TCP acceptor changed a bit based on Aditya's suggestions
* Updated the example.yaml comments based on Cory's suggestions.

Let me know what you think, and thanks again!

evanj commented:
I just rebased this change again, since there were a bunch of conflicts with recent changes. Looks like it passed the Travis tests, so I'd love a re-review when you get a chance. Thanks!

ChimeraCoder commented on server_test.go:
```diff
@@ -569,3 +584,262 @@ func assertCSVFieldsMatch(t *testing.T, expected, actual [][]string, columns []i
+	}
+}
+
+// DO NOT USE: insecure test keys generated with the following commands:
```
Can we either put these in a separate `.go` file (rather than `server_test.go`) or in a fixture (like we do for the JSON examples? My preference is for the latter, since it's more consistent, but either one works. 

If we use a fixture, we could read the certs in the `init()` for the tests and include this comment there.

evanj replied:
I just added you as a collaborator if you would prefer to do this yourself. I'm also happy to take care of it. If you want me to take care of it: You want me to add these as separate files in the `fixtures` directory? I'm happy to give that a shot.

ChimeraCoder replied:
Yup! Just add them as `.pem` files in the `fixtures` and read them on test startup.

I'm heading home in a little bit, and I won't realistically have time until tomorrow afternoon myself either. If you haven't had a chance to do it by then, I'll do it myself before merging - no worries either way.

evanj commented:
I made an attempt at this. Let me know what you think, or feel free to just directly change what you don't like about this. I won't be offended. Thanks again for all the help!

evanj commented:
Hello! I noticed this has been labelled "approved" but is still not merged. Is there anything else do be done here?

cory-stripe commented:
Great questions! Forgive our latency, @aditya-stripe is traveling this week and I'm tasked with a couple distractions. I'll give this a once over again and merge before EoD tomorrow it if nothing comes up.

Thanks again for the patience @evanj.

gphat commented:
Thanks again @evanj! I'll merge this now and update the changelog.

gphat commented:
Nevermind, you added it already! Again, thanks for the patience and thoroughness here @evanj!

evanj commented:
VICTORY Thanks! Actually @ChimeraCoder added the Changelog entry.

I'll send you an email with additional details, but thanks so much! This is in production at Bluecore to aggregate custom metrics from our App Engine app, at a peak rate of about ~10000 metrics/second and a peak utilization about about ~50% of one core. That tends to happen when a bunch of connections close and reopen, since the TLS negotiation is relatively expensive.

cory-stripe commented:
Neat! Glad to hear it's working well and even more glad to've added a new feature in the process! :)

wchampi commented:
Hello 

Could you help me? How can I config Datadog Agent to forward statsd metrics using TLS?

ChimeraCoder commented:
@wchampi This patch simply enables Veneur to accept data from trusted client certificates. Datadog Agent is a separate project maintained by Datadog. If you're having issues with with configuring [TLS for Datadog Agent using a proxy](https://docs.datadoghq.com/agent/proxy/?tab=agentv6v7), I'd suggest contact them for support.
